### PR TITLE
batches: allow batch change `last_applied_at` to be nullable

### PR DESCRIPTION
**Pull request by courier-new**
Part 1 of EM2.4 (https://github.com/sourcegraph/sourcegraph/issues/26796), which adds the DB migration to allow `last_applied_at` on the batch change to be nullable, as it would be for an empty batch change.

For the down migration, I chose to set the column value for empty batch changes to the "zero" timestamp aka the unix epoch time as the closest approximation to not having an applied-at date, but realistically this isn't a change we should anticipate needing to revert for any instance that has empty batch changes active on it.

### CODE REVIEW

File: migrations/frontend/1528395945_make_batch_change_last_applied_at_nullable.up.sql [hunk 1]
```diff
@@ -0,0 +1,11 @@
+BEGIN;
+
+ALTER TABLE IF EXISTS batch_changes
+ALTER COLUMN last_applied_at
+DROP NOT NULL;
+
+ALTER TABLE IF EXISTS batch_changes
+ALTER COLUMN last_applied_at
+DROP DEFAULT;
```
Comment (mrnugget) [inline migrations/frontend/1528395945_make_batch_change_last_applied_at_nullable.up.sql:9]: Nitpick: I'd personally try to reduce the amount of text to make it easier to spot what the important bit is and combine those two statements

```suggestion
ALTER TABLE IF EXISTS batch_changes
  ALTER COLUMN last_applied_at DROP NOT NULL,
  ALTER COLUMN last_applied_at DROP DEFAULT;
```

File: migrations/frontend/1528395945_make_batch_change_last_applied_at_nullable.down.sql [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+UPDATE batch_changes SET last_applied_at = TO_TIMESTAMP(0) WHERE last_applied_at IS NULL;
+
+ALTER TABLE IF EXISTS batch_changes
+ALTER COLUMN last_applied_at
+SET DEFAULT NOW();
+
+ALTER TABLE IF EXISTS batch_changes
+ALTER COLUMN last_applied_at
+SET NOT NULL;
```
Comment (mrnugget) [inline migrations/frontend/1528395945_make_batch_change_last_applied_at_nullable.down.sql:11]: Same as below.

**Review by mrnugget — APPROVED** [review]
LGTM!

Just as a note (not blocking): I'm wondering whether this PR isn't too small. The `batches/store.Store` still assumes that the field is non-nullable. Yes, in practice this won't be a problem, yes, we also have other PRs that only work when merged together, but I just wanted to mention it. I personally would change database constraints + store always in a single PR so that reverting and or migrating up/down works better.

**Review by eseliger — APPROVED** [review]
Agree with Thorsten's comment that it feels a bit safer when all of the store changes at least are done together.

File: migrations/frontend/1528395945_make_batch_change_last_applied_at_nullable.up.sql (hunk 1 above)
Comment (courier-new) [reply migrations/frontend/1528395945_make_batch_change_last_applied_at_nullable.up.sql:9]: Great suggestion, thank you!

**Comment by courier-new** [comment]
Changing the store at the same time as these DB migrations makes sense to me, thank you both for raising that! 😄

**Pull request merged**